Add Gazell support (#5281)
nrfconnect/sdk-nrf: doc/nrf/ug_gzll.rst

Gazell automatically handles all synchronization and packet handling.
All the user is required to do is to add payloads to the transmit (TX) FIFOs and read payloads from the receive (RX) FIFOs.
Gazell automatically notifies the application when a packet is received.

To set up a Gazell application, do the following:

* Initialize Gazell using nrf_gzll_init() and choose either Host or Device.
* Reconfigure Gazell's default parameters.
  At a minimum the addresses and channels should be reconfigured to avoid interfering with other Gazell networks.
* Enable Gazell using nrf_gzll_enable().
* If the node is a Device, start sending:

  * Add payloads to the TX FIFO using nrf_gzll_add_packet_to_tx_fifo().
  * Handle the returned ACK packet when the nrf_gzll_device_tx_success() callback is called.
    Fetch the payloads from the RX FIFO using nrf_gzll_fetch_packet_from_rx_fifo().
  * Handle the failed packet transmissions when the nrf_gzll_device_tx_failed() callback is called.
    Failed packets are automatically removed from the TX FIFO.

* If the node is a Host, start listening:

  * Handle the received data packets when the nrf_gzll_host_rx_data_ready() calbback is called.

>     Fetch the packets from the RX FIFO using nrf_gzll_fetch_from_rx_firo().

Author: It is a typo. It should be 'fifo'.

>     Fetch the packets from the RX FIFO using nrf_gzll_fetch_from_rx_firo().

Author: Will fix it.

This prevents the application receiving duplicate data packets at the Host.

.. figure:: images/gzll_fig8_host_dev_trans_fail.svg
   :alt: Example on failing package transaction.

   Example on failing package transaction.

In the figure above the maximum number of allowed transmission attempts is set to 3.

Packet identification
*********************

Any packet transmitted from a Device to a Host is uniquely identified by a two bit packet ID field in the packet header together with the packet's 16-bit Cyclic Redundancy Check (CRC).
This packet ID is used to distinguish a new packet from the previous packet if it has the same payload.

On the Host side, retransmitted packets will be discarded and not added to an RX FIFO.

Pipes and addressing
====================

Each logical address on the nodes is termed a pipe.
Each pipe maps to one on-air address used when transmitting or receiving packets.

The on-air addresses are composed of a 2-4 byte long "base address" in addition to a 1 byte prefix address.
Note that the nRF5 radio uses an alternating sequence of 0s and 1s as the preamble of the packet.

> Therefore, for packets to be received correctly, the most significant byte of the base address should not be an alternating sequence of 0s and 1s, that is, it should not be 0x55 or 0xAA.

Reviewer: If you want this to stand out from the text, please use the note formatting like:

.. note:
   The nRF5 radio...

Pipe 0 has its own unique base address, which is base address 0, while pipes 1-7 use the same base address, which is base address 1.

> Each of the 8 pipes have a unique byte-long prefix address.

Reviewer: ```suggestion
Each of the eight pipes have a unique byte-long prefix address.
```

On-air, the most significant bit of each address byte will be transmitted first.

> The most significant byte of the 4 byte long base address is the first transmitted address byte, while the prefix byte is transmitted last.

Reviewer: ```suggestion
The most significant byte of the four bytes long base address is the first transmitted address byte, while the prefix byte is transmitted last.
```

> Note: The byte ordering in Gazell and the nRF5 radio peripheral are not the same, this is because the address bytes are rearranged in Gazell to match the nRF24L radios.

Reviewer: ```suggestion
.. note: 
   The byte ordering in Gazell and the nRF5 radio peripheral are not the same.
    This is because the address bytes in Gazell are rearranged to match the nRF24L radios.
```

Features
********

Gazell provides the following features:

* Support for star network topology with one Host and up to eight Devices.
* Bidirectional data transfer between each Host and Device.
* Channel hopping functionality that gives a reliable wireless link in environments with interference from other radio sources.
* Packet acknowledgment and automatic packet retransmission functionality to prevent data loss.
* Individual TX and RX FIFOs for every data pipe.
* Backward compatible with legacy nRF24L Series Gazell.
* Devices self-synchronize to the Host, meaning:

  * No connection packets are required to setup a link.
  * No polling packets are required to maintain a link.
  * Devices can enter and remove themselves from the network at any time.

* Capability to control the RF front-end module.
* Generates transmission statistics for each RF channel.

.. _ug_gzll_resources:

Resources
*********

Gazell uses a fixed set of peripheral resources in the nRF5 SoC.
To ensure correct operation, Gazell requires exclusive access to the following resources:

* Radio
* Timer

> * 3 PPI channels

Reviewer: ```suggestion
* Three PPI channels
```

Author: Will do.

Each Device communicates to a single Host.

.. figure:: images/gzll_fig1_star_network.svg
   :alt: Gazell star network

   Gazell star network

Once enabled, the Host in a Gazell network is always listening, and the Device always initiates the communication.
Each packet that a Device sends is required to be acknowledged by the Host.
The Host can send data to the Device by piggybacking data in an acknowledgment (ACK) packet.
Therefore, the Host must wait for a packet from the Device before it can send any data to it.

You can build more sophisticated Gazell networks, since a single Device can speak to several Hosts and any node can change between the two modes.
However, this requires the application to coordinate such a network.

This document focuses on the typical use case of a star network with static modes.

Setting up a Gazell application
*******************************

Gazell automatically takes care of all synchronization and packet handling.
You need to add payloads to the transmit (TX) FIFOs and read payloads from the receive (RX) FIFOs.
Gazell automatically notifies the application when a packet is received.

To set up a Gazell application, do the following:

> * Initialize Gazell using ``nrf_gzll_init()`` and choose either Host or Device.

Reviewer: ```suggestion
* Initialize Gazell using :c:func:`nrf_gzll_init()` and choose either Host or Device.
```
Please, use the same formatting for all functions.

In Device mode, ``timeslots_per_channel`` can also be seen as the number of transmission attempts spent on each channel before switching the channel.
This is because there is at least one transmission attempt for every timeslot.

The ``channel_selection_policy`` parameter is used by a Device in sync to decide the initial channel to be used when sending a new packet to a Host (that is, for the first time the new packet is sent, not for the retransmission attempts).

Once synchronized with the Host, the Device can send either on the current channel that it believes the Host is on or on the last successful channel.
To configure this, use the ``nrf_gzll_set_device_channel_selection_policy()`` parameter.

The ``channel_selection_policy`` parameter can take the following two values:

* NRF_GZLL_DEVICE_CHANNEL_SELECTION_POLICY_USE_SUCCESSFUL
* NRF_GZLL_DEVICE_CHANNEL_SELECTION_POLICY_USE_CURRENT

If you choose the NRF_GZLL_DEVICE_CHANNEL_SELECTION_POLICY_USE_SUCCESSFUL policy, the Device starts sending packets on the channel it last had a successfully acknowledged transmission.
This policy is the most robust against static interference.

> Once the Device finds a quiet channel it should be able to continue using this channel.

Reviewer: ```suggestion
Once the Device finds a quiet channel, it should be able to continue using this channel.
```

If you choose the NRF_GZLL_DEVICE_CHANNEL_SELECTION_POLICY_USE_CURRENT policy, the Device sends on the channel it believes the Host is currently listening to.
This achieves the lowest latency and highest throughput of the two policies as the Device does not have to wait for the Host to be listening to a specific channel.
This policy is frequency hopping.
The disadvantage of this policy is that if there is static interference on a particular channel, the Device wastes packets attempting to send on this channel.
The application can reconfigure the channel table during runtime to overcome this.

The channel selection policy only applies to the initially transmitted packet.
If the transmission of this initial packet fails, the following retransmission attempts are always sent in the channel the Device believes the Host is monitoring.

If Gazell is "out of sync", it always starts the packet transmission immediately using the previous successful transmission channel.

> If Gazell has not transmitted a successful packet and thus has no "previous successful channel" to relate to, it starts using the first channel in the channel table.

Reviewer: ```suggestion
If Gazell has not transmitted a successful packet and thus has no previous successful channel to relate to, it starts using the first channel in the channel table.
```

.. _gazell_sync:

Synchronization
***************

The internal timeslot, or "heartbeat", mechanism of Gazell is used to obtain synchronous communication while still enabling efficient channel switching.

> This mechanism is useful when a Device needs to switch to a new channel when there is radio interference on the current channel.

Reviewer: ```suggestion
This mechanism is useful when a Device needs to switch to a new channel while there is radio interference on the current channel.
```

> Each Gazell Device has two synchronization states: "in sync" and "out of sync".

Reviewer: ```suggestion
Each Gazell Device has two synchronization states: in sync and out of sync.
```

> On the Host, the internal "heartbeat" timer is always running when Gazell is enabled, independent of the Devices' synchronization state.

Reviewer: ```suggestion
On the Host, the internal heartbeat timer is always running when Gazell is enabled, independent of the Devices' synchronization state.
```

> On the Device, the "heartbeat" timer only runs as long as the Device is "in sync" or as long as there are packets to be sent.

Reviewer: ```suggestion
On the Device, the heartbeat timer only runs as long as the Device is in sync or as long as there are packets to be sent.
```

If the timer has been stopped and packets are added to a TX FIFO, the timer starts immediately.

Before any packets have been successfully received and acknowledged, the Device is out of sync.
In this state, the Device switches channel determined by the ``timeslots_per_channel_when_device_out_of_sync`` parameter.
The Device switches channel at a slower rate than the Host (as determined by ``timeslots_per_channel``) so that the Device eventually transmits a packet on the same channel that the Host is on.

> When a Device successfully transmits a packet, that is when an ACK packet is received from the Host, it enters the "in sync" state, as it now has the information needed for continuing to "guess" the following channels the Host is listening to.

Reviewer: ```suggestion
When a Device successfully transmits a packet, that is when an ACK packet is received from the Host, it enters the in sync state, as it now has the information needed for continuing to guess the following channels the Host is listening to.
```

For knowing when to change channel, Gazell has an internal ``timeslot_counter`` to count the number of timeslots it has on a single channel.

> When this counter reaches ``timeslots_per_channel``, the ``timeslot_counter`` is reset and the channel_index is incremented (cyclically).

Reviewer: ```suggestion
When this counter reaches ``timeslots_per_channel``, the ``timeslot_counter`` is reset and the ``channel_index`` is incremented (cyclically).
```

> When an ACK is received, the Device knows the Host is using the current channel, but it does not know the ``timeslot_counter`` state on the Host.

Reviewer: ```suggestion
When the Device has received an ACK, it knows the Host is using the current channel, but it does not know the ``timeslot_counter`` state on the Host.
```

As a result, only in the timeslots where the ``timeslot_counter`` equals zero the Device can be confident that it "guesses" the correct channel that the Host is monitoring.
Therefore, when an ACK is received, the ``timeslot_counter`` for the current timeslot is reset to zero, and a new Device transmission starts when the ``timeslot_index`` counter on the Device is zero.
Retransmission attempts, however, are sent on all timeslots.

Once the Device is in sync, it keeps an internal timer running to maintain the internal heartbeat in order to remain synchronized with the Host.
The duration the Device stays in the in sync state is the ``sync_lifetime`` and is measured in timeslots.
The ``sync_lifetime`` is reset whenever a packet is received.
Once the ``sync_lifetime`` has expired on a Device, the internal timer is stopped and the Device returns to out of sync state.

.. note:

>    When a Device that is "in sync" sends a packet but does not receive an ACK, it continues to transmit until it reaches the maximum number of attempts.

Reviewer: ```suggestion
   When a Device that is in sync sends a packet but does not receive an ACK, it continues to transmit until it reaches the maximum number of attempts.
```

If you set the ``sync_lifetime`` to zero, the Device will never be in sync.
The ``sync_lifetime`` should be chosen with regard to how often packets are required to be sent and the fact that synchronization can only be maintained for a finite time due to clock drift and radio interference.

> To configure the sync lifetime use the ``nrf_gzll_set_sync_lifetime()`` parameter.

Reviewer: ```suggestion
To configure the sync lifetime, use the ``nrf_gzll_set_sync_lifetime()`` parameter.
```

The Device knows the it is in sync when the number of retransmissions gets close to zero.
The ``nrf_gzll_device_tx_info_t`` structure is passed to the Device callback functions, and it contains the number of transmit attempts required for the current packet.
In addition, the structure contains the ``num_channel_switches`` parameter that the application can use to determine whether the RF channels are reliable.
This enables the application to track bad channels and update the channel tables on Host and Device if desired.

Backwards compatibility
***********************

The Gazell Link Layer examples are not fully "out of the box" compatible with the legacy examples provided in the nRFgo SDK for nRF24Lxx devices.

> The default timeslot period and channel tables require adjustment, as well some setup to emulate the Gazell modes.

Reviewer: ```suggestion
The default timeslot period and channel tables require adjustment, as well as some setup to emulate the Gazell modes.
```

> The Gazell "Low Power Host mode" (Host mode 1) is not supported in the nRF5.

Reviewer: ```suggestion
The Gazell Low Power Host mode (Host mode 1) is not supported in the nRF5.
```

Channel tables
==============

The default channel tables require adjustment.

> Depending on your project, one of the following:

Reviewer: ```suggestion
Depending on your project, do one of the following:
```

* Edit the :file:`gzll_params.h` file used in the nRF24Lxx projects.
* Use the ``nrf_gzll_set_channel_table()`` function in the nRF5 projects.

Timeslot periods
================

The Gazell Link Layer supports the following minimum timeslot periods:

* 600 us timeslot period, nRF5 Gazell Device to nRF5 Gazell Host.
* 504 us timeslot period, nRF5 Gazell Device to nRF24Lxx Gazell Host.

When using 504 us timeslot period, the following restrictions apply:

> * Max payload size is 17 bytes.
> * Max ack payload size is 10 bytes.

Reviewer: ```suggestion
* The maximum payload size is 17 bytes.
* The maximum ack payload size is 10 bytes.
```

In addition, the relation between the Device and Host timing parameters should be as follows:

* The Host listens to each channel in a GZLL_RX_PERIOD number of microseconds, where GZLL_RX_PERIOD is the heartbeat interval in the nRF24Lxx devices.

> * The Host GZLL_RX_PERIOD must be greater than the time required committing two full transmission attempts on the Device (including ACK wait time).

Reviewer: ```suggestion
* The Host GZLL_RX_PERIOD must be greater than the time required to make two full transmission attempts on the Device (including ACK wait time).
```

> Depending on your project, one of the following:

Reviewer: ```suggestion
Depending on your project, do one of the following:
```

* Edit the :file:`gzll_params.h` file used in the nRF24Lxx projects.
* Use the ``nrf_gzll_set_timeslot_period()`` function in the nRF5 projects (nRF5 Gazell timeslot period = 0.5*GZLL_RX_PERIOD).

Emulating legacy Gazell modes
=============================

The Gazell Link Layer protocol for the nRF5 Series is compatible with the most useful modes of the Gazell Link Layer for the nRF24Lxx devices.

Emulating legacy nRF24Lxx Gazell Device mode 2 and nRF24Lxx Host mode 0
-----------------------------------------------------------------------

> You can emulate the legacy "Device mode 2" as follows:

Reviewer: ```suggestion
You can emulate the legacy Device mode 2 as follows:
```

* The channel selection policy is equivalent to NRF_GZLL_DEVICE_CHANNEL_SELECTION_POLICY_USE_SUCCESSFUL.

> * When Gazell is "out of sync", a large number of attempts may occur on each channel before the channel is switched.

Reviewer: ```suggestion
* When Gazell is out of sync, a large number of attempts may occur on each channel before the channel is switched.
```

> * When Gazell is "in sync", a low number of transmission attempts, typically two, are allowed on each channel before the channel is switched.

Reviewer: ```suggestion
* When Gazell is in sync, a low number of transmission attempts, typically two, are allowed on each channel before the channel is switched.
```

> The legacy "Host mode 0" behaves as follows:

Reviewer: ```suggestion
The legacy Host mode 0 behaves as follows:
```

* Host is always on while it is enabled.
* When enabled, the Host will continuously cycle through the channel table.

See the example on how to achieve such behavior.

> Assuming a channel table ``my_channel_table[]`` with 3 channels.

Reviewer: ```suggestion
Assuming a channel table ``my_channel_table[]`` with three channels:
```

   nrf_gzll_set_device_channel_selection_policy(NRF_GZLL_DEVICE_CHANNEL_SELECTION_POLICY_USE_SUCCESSFUL);

.. figure:: images/gzll_fig9_gzll_config_example.svg
   :alt: Emulating legacy Gazell

   Emulating legacy Gazell

Transmission statistics
***********************

The Gazell stack allows to automatically gather transmission information, such as:

* Total number of transmitted packets.
* Total number of transmission timeouts.
* Number of transmitted packets on each RF channel.
* Number of transmission failures on each RF channel.

The stack can also track packet transaction failure events, such as transmission timeout or receiving a packet with incorrect CRC.

To turn on transmission statistics, perform the following steps:

1. Define the ``nrf_gzll_tx_statistics_t`` structure.
   This is a buffer for transmission statistics data, so it must remain in memory as long as the transmission statistics are used.
#. Call ``nrf_gzll_init`` to initialize Gazell.
#. Call the ``nrf_gzll_tx_statistics_enable`` function to enable transmission information gathering.

After this, transmission statistics can be read from the defined structure.

> To reset recording, call the ``nrf_gzll_reset_tx_statistics`` function.

Reviewer: ```suggestion
To reset the recording, call the ``nrf_gzll_reset_tx_statistics`` function.
```
